### ISSUE TITLE
Accessing object property

### ISSUE THREAD

**Issue by marcpearson** [alonrbar/easy-template-x#39]
My data (received from backend API) is organized as hierarchical content. How can access an object property within a template file

Something like {project.client.title}

I found no way to make it work

**Comment by alonrbar** [comment]
**TL;DR**
The library does not support nested properties. You cannot use your data from your backend API as-is, instead you will have to create a modified "template data" object to pass to your TemplateHandler instance.

**A little bit of philosophy...**
Think of your data in the context of the template and structure your JSON the way the template is structured, and not the other way around. Depending on how your template looks like you should structure your data accordingly. What are you rendering? If for instance you are rendering an array of projects where each has a title your template may look like this:
```
{#projects}
{title}
{/projects}
```
And your data like this:
```json
{
  "projects": [
    { "title": "first project" },
    { "title": "second project" }
  ]
}
```
Of course, that's just an example. If you need more help designing your template and "template data" I'll be happy to help if you'll provide more context and elaborate more on your requirements.

**Comment by tom-acceleratecompliance** [comment]
Thanks @alonrbar for the fantastic plugin, and apologies for going against the philosophy, but here is what I did:

```javascript
const nestedData = { foo: { bar: 'Widget' } } 

// Setup plugin...

handler.process(template, unnest(nestedData))

function unnest(obj) {
    var res = {};
    (function recurse(obj, current) {
      for (var key in obj) {
        var value = obj[key];
        var newKey = (current ? current + "." + key : key);  // joined key with dot
        if (value && typeof value === "object") {
          recurse(value, newKey);  // it's a nested object, so do it again
        } else {
          res[newKey] = value;  // it's not an object, so set the property
        }
      }
   })(docData);
  return res
}

```

This effectively creates a flat object: `{ 'foo.bar': 'Widget' }` which works in the template with `{foo.bar}`
There may be some edge cases I have not considered though.

**Comment by alonrbar** [comment]
Hi @tom-acceleratecompliance,
No problem at all :) The "philosophy" part is just my way of thinking on the problem and by all means users are welcome to do what best works for them. This looks like a nice solution that can potentially help other users who prefer it this way and I invite you to publish an [extension](https://github.com/alonrbar/easy-template-x#extensions) for your solution. If you chose to do so I'll be happy to post a link to it in the readme.

**Comment by tom-acceleratecompliance** [comment]
Hi @alonrbar - thanks I'll take a look.
Just as an update, the above code doesn't account for arrays, see updated code:

```javascript
  unnest(docData) {
    var res = {};
    (function recurse(obj, current) {
      for (var key in obj) {
        var value = obj[key];
        var newKey = (current ? current + "." + key : key);  // joined key with dot
        if (value && typeof value === "object") {
          if (Array.isArray(value)) {
            res[newKey] = value.map(function(e) { 
              if(typeof e == 'object') {
                return unnest(e)
              } else {
                return e
              }
            })
          } else {
            recurse(value, newKey);  // it's a nested object, so do it again
          }
        } else {
          res[newKey] = value;  // it's not an object, so set the property
        }
      }
    })(docData);
    return res
  }
```

**Comment by alonrbar** [comment]
Closing this issue.

If an extension that support this behavior is published let me know and I'll add a link to it in the readme.

**Comment by alonrbar** [comment]
UPDATE: This is now possible with [easy-template-x-angular-expressions](https://github.com/alonrbar/easy-template-x-angular-expressions).

**Comment by bitfabrikken** [comment]
@alonrbar can you explain how to unnest with the angular expressions extension? 
Or maybe it allows accessing {foor.bar.something} somehow? 
I can submit a pull request to the README.md with an example.

**Comment by alonrbar** [comment]
Yes, exactly it allows exactly that. Please see the docs there.
